iplass-core/src/main/java/org/iplass/mtp/impl/auth/authenticate/token/CachableAuthTokenStore.java

```diff
@@ -0,0 +1,263 @@
+					tokenCache.notifyDeleteByKey(key);
+					keys.add(key);
+				}
+			}
+			
+			//念のため
+			Transaction t = Transaction.getCurrent();
+			if (t.getStatus() == TransactionStatus.ACTIVE) {
+				t.afterCommit(() -> {
+					for (AuthTokenKey k: keys) {
+						tokenCache.notifyInvalidByKey(k);
+					}
+				});
+			}
+		}
+	}
+
+	@Override
+	public void deleteBySeries(int tenantId, String type, String series) {
+		authTokenStore.deleteBySeries(tenantId, type, series);
+		AuthTokenKey key = new AuthTokenKey(tenantId, type, series);
+		tokenCache.notifyDeleteByKey(key);
+		
+		//念のため
+		Transaction t = Transaction.getCurrent();
+		if (t.getStatus() == TransactionStatus.ACTIVE) {
+			t.afterCommit(() -> {
+				tokenCache.notifyInvalidByKey(key);
+			});
+		}
+	}
+
+	@Override
+	public void deleteByDate(int tenantId, String type, Timestamp ts) {
+		//キャッシュの有効期間は、ここで指定される日付より十分短い想定
+		authTokenStore.deleteByDate(tenantId, type, ts);
+	}
+
+	private static class AuthTokenKey implements Serializable {
+		private static final long serialVersionUID = 6033114232648174950L;
+
+		private final int tenantId;
+		private final String type;
+		private final String series;
+		
+		public AuthTokenKey(int tenantId, String type, String series) {
+			this.tenantId = tenantId;
+			this.type = type;
+			this.series = series;
+		}
+
+		public int getTenantId() {
+			return tenantId;
+		}
+
+		public String getType() {
+			return type;
+		}
+
+		public String getSeries() {
+			return series;
+		}
+
+		@Override
+		public int hashCode() {
+			final int prime = 31;
+			int result = 1;
+			result = prime * result + ((series == null) ? 0 : series.hashCode());
+			result = prime * result + tenantId;
+			result = prime * result + ((type == null) ? 0 : type.hashCode());
+			return result;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj)
+				return true;
+			if (obj == null)
+				return false;
+			if (getClass() != obj.getClass())
+				return false;
+			AuthTokenKey other = (AuthTokenKey) obj;
+			if (series == null) {
+				if (other.series != null)
+					return false;
+			} else if (!series.equals(other.series))
+				return false;
+			if (tenantId != other.tenantId)
+				return false;
+			if (type == null) {
+				if (other.type != null)
+					return false;
+			} else if (!type.equals(other.type))
+				return false;
+			return true;
+		}
+
+	}
+
+	public static class AuthTokenKeyResolver implements CacheKeyResolver {
+
+		@Override
+		public String toString(Object cacheKey) {
+			AuthTokenKey key = (AuthTokenKey) cacheKey;
+			StringBuilder sb = new StringBuilder();
+			sb.append(key.getTenantId()).append(':').append(key.getType()).append(':').append(key.getSeries());
+			return sb.toString();
+		}
+
+		@Override
+		public Object toCacheKey(String cacheKeyString) {
+			int index = cacheKeyString.indexOf(':');
+			int tenantId = Integer.parseInt(cacheKeyString.substring(0, index));
+			int index2 = cacheKeyString.indexOf(index + 1, ':');

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

If the indexOf() search on line 254 does not find the substring and returns -1, the call to substring() on line 255 will throw IndexOutOfBoundsException.